<h1 style="color:#1E90FF; font-weight:bold;">BTS</h1>
<h2 style="color:#1E90FF; font-weight:bold;">MBDS - Big Data & AI</h2>
<h3 style="color:#1E90FF; font-weight:bold;">DSF - Assignment 1</h3>
<h4 style="color:#1E90FF; font-weight:bold;">Pablo Ruiz Lopez</h4>

In this assignment, my primary approach to answering the questions will leverage the pandas library in Python. However, for exploratory, curiosity, experimentation and comparative purposes, I will also employ the [polars python library](https://blog.jetbrains.com/dataspell/2023/08/polars-vs-pandas-what-s-the-difference/#:~:text=Pandas%2C%20by%20default%2C%20uses%20eager,way%20of%20executing%20the%20code.) to provide alternative solutions. 

**NOTE:** As part of this comparisson, you will see the magic command `%%time` at the beginning of each cell in order to evaluate differences in execution times between solutions.

_But why consider using [polars](https://www.pola.rs/)?_

**Performance:** Superior performance compared to pandas, especially when dealing with large datasets. Crafted in Rust and optimized for parallel execution, polars ensures rapid operations, enhancing efficiency.

**Memory Efficiency:** Polars is recognized for its memory optimization. It utilizes resources effectively, making it a preferable option for tasks involving substantial data volumes.

**Functionality:** Eventhough polars is still in its nascent stages. While it aspires to mirror the functionality of pandas, it may not yet boast the comprehensive features or the extensive community and third-party support that pandas enjoys.

**Compatibility:** Interoperability between polars and pandas is another advantage. We can convert DataFrames between the two libraries. This flexibility is instrumental when looking to harness polars for intensive computations while relying on pandas for specific, nuanced features.

**In Essence:**

The decision to utilize either [polars or pandas](https://medium.com/cuenex/pandas-2-0-vs-polars-the-ultimate-battle-a378eb75d6d1) hinges on distinct project requisites, such as performance imperatives, dataset dimensions, or the affinity for a specific API. Both libraries are tailored for intricate data manipulation and analytical tasks, each presenting unique advantages that cater to diverse computational and analytical needs.

<div style="text-align: center;">
    <img src="https://images.datacamp.com/image/upload/v1686564840/datarhys_an_absurdist_oil_painting_of_a_cyborg_panda_and_a_cybo_c7deba5e_79f3_498a_ada9_381d6fe0b77a_b943fa043f.png" width="500">
</div>

<div style="text-align: center;">
    <a href="https://www.datacamp.com/tutorial/high-performance-data-manipulation-in-python-pandas2-vs-polars">Click HERE for a great library comparison</a>
</div>

In [1]:
# Imports

import pandas as pd
import polars as pl

In [2]:
# Loading the dataset into a pandas DataFrame
# Using pandas to load the data first due to an error in Polars when parsing the 'Codi_Districte' column as an integer
df_pd = pd.read_csv('2020_antiguitat_tipus_vehicle.csv')

# Converting the pandas DataFrame to a polars DataFrame
# Work around for the parsing issue
df_pl = pl.from_pandas(df_pd)

#### **Basic Data Exploration**

In [3]:
# Information about the columns, data types and values
df_pd.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52977 entries, 0 to 52976
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Any             52977 non-null  int64 
 1   Codi_Districte  52977 non-null  object
 2   Nom_Districte   52977 non-null  object
 3   Codi_Barri      52977 non-null  object
 4   Nom_Barri       52977 non-null  object
 5   Seccio_Censal   52977 non-null  object
 6   Tipus_Vehicles  52977 non-null  object
 7   Antiguitat      52977 non-null  object
 8   Nombre          52977 non-null  int64 
dtypes: int64(2), object(7)
memory usage: 3.6+ MB


In [4]:
# Displaying the 5 first records using polars
df_pl.head()

shape: (5, 9)
┌──────┬─────────────┬────────────┬────────────┬───┬────────────┬────────────┬────────────┬────────┐
│ Any  ┆ Codi_Distri ┆ Nom_Distri ┆ Codi_Barri ┆ … ┆ Seccio_Cen ┆ Tipus_Vehi ┆ Antiguitat ┆ Nombre │
│ ---  ┆ cte         ┆ cte        ┆ ---        ┆   ┆ sal        ┆ cles       ┆ ---        ┆ ---    │
│ i64  ┆ ---         ┆ ---        ┆ str        ┆   ┆ ---        ┆ ---        ┆ str        ┆ i64    │
│      ┆ str         ┆ str        ┆            ┆   ┆ str        ┆ str        ┆            ┆        │
╞══════╪═════════════╪════════════╪════════════╪═══╪════════════╪════════════╪════════════╪════════╡
│ 2020 ┆ 1           ┆ Ciutat     ┆ 1          ┆ … ┆ 1          ┆ Turismes   ┆ Menys d'un ┆ 11     │
│      ┆             ┆ Vella      ┆            ┆   ┆            ┆            ┆ any d'anti ┆        │
│      ┆             ┆            ┆            ┆   ┆            ┆            ┆ guitat     ┆        │
│ 2020 ┆ 1           ┆ Ciutat     ┆ 1          ┆ … ┆ 1          ┆ Turismes   ┆ 1 any      ┆ 23     │
│      ┆             ┆ Vella      ┆            ┆   ┆            ┆            ┆            ┆        │
│ 2020 ┆ 1           ┆ Ciutat     ┆ 1          ┆ … ┆ 1          ┆ Turismes   ┆ 2 anys     ┆ 11     │
│      ┆             ┆ Vella      ┆            ┆   ┆            ┆            ┆            ┆        │
│ 2020 ┆ 1           ┆ Ciutat     ┆ 1          ┆ … ┆ 1          ┆ Turismes   ┆ 3 anys     ┆ 18     │
│      ┆             ┆ Vella      ┆            ┆   ┆            ┆            ┆            ┆        │
│ 2020 ┆ 1           ┆ Ciutat     ┆ 1          ┆ … ┆ 1          ┆ Turismes   ┆ 4 anys     ┆ 17     │
│      ┆             ┆ Vella      ┆            ┆   ┆            ┆            ┆            ┆        │
└──────┴─────────────┴────────────┴────────────┴───┴────────────┴────────────┴────────────┴────────┘

In [5]:
# The NC value is the one non-parsable as int64
df_pl['Codi_Districte'].unique()

shape: (11,)
Series: 'Codi_Districte' [str]
[
	"6"
	"1"
	"9"
	"3"
	"NC"
	"7"
	"5"
	"10"
	"4"
	"8"
	"2"
]

### **Question #1**

What is the total number of vehicles in Barcelona? (number of cars is in field Nombre) (10 points)


In [6]:
%%time
# Pandas solution

# Computing the sum of 'Nombre' 
df_pd['Nombre'].sum()

Wall time: 0 ns


822211

In [7]:
%%time 
# Polars solution

# Computing the sum of 'Nombre' 
df_pl['Nombre'].sum()

Wall time: 0 ns


822211

**A:** There are **822,211** vehicles in Barcelona registered in 2020. 

### **Question #2**

What is the the number of vehicles by type (Tipus_Vehicles). 2 columns, 'Tipus_Vehicles' and the total vehicles for that type. (hint: groupby) (10 points)

In [8]:
%%time 
# Pandas solution

# Grouping by type of vehicle and computing the sum of 'Nombre'
df_pd.groupby('Tipus_Vehicles', as_index = False)['Nombre'] \
        .sum()

Wall time: 8.84 ms


Tipus_Vehicles  Nombre
0  Altres vehicles   17285
1          Camions   15344
2      Ciclomotors   51536
3       Furgonetes   39177
4            Motos  219554
5         Turismes  479315

In [9]:
%%time 
# Polars solution

# Grouping by type of vehicle and computing the sum of 'Nombre' using aggregation function
df_pl.group_by('Tipus_Vehicles') \
        .agg(pl.col('Nombre').sum())

Wall time: 43.5 ms


shape: (6, 2)
┌─────────────────┬────────┐
│ Tipus_Vehicles  ┆ Nombre │
│ ---             ┆ ---    │
│ str             ┆ i64    │
╞═════════════════╪════════╡
│ Camions         ┆ 15344  │
│ Ciclomotors     ┆ 51536  │
│ Altres vehicles ┆ 17285  │
│ Turismes        ┆ 479315 │
│ Furgonetes      ┆ 39177  │
│ Motos           ┆ 219554 │
└─────────────────┴────────┘

### **Question #3**
How many vehicles (any type) with age (Antiguitat) >= 10 years do we have in district ('Nom_Districte') Eixample? (10 points)

In [10]:
# Unique values included in 'Antiguitat' column
df_pl['Antiguitat'].unique()

shape: (14,)
Series: 'Antiguitat' [str]
[
	"Menys d'un any…
	"2 anys"
	"1 any"
	"7 anys"
	"5 anys"
	"10 anys"
	"3 anys"
	"Més de 20 anys…
	"8 anys"
	"D'11 a 20 anys…
	"6 anys"
	"9 anys"
	"4 anys"
	"No consta"
]

`Antiguitat` values depicting 10 or more years:

* "10 anys": 10 years
* "D'11 a 20 anys": from 11 to 20 years
* "Més de 20 anys": more than 20 years

In [11]:
%%time
# Pandas solution

# Filtering based on 'Antiguitat' and 'Nom_Districte'
df_pd[
      (df_pd['Antiguitat'].isin(["10 anys", "D'11 a 20 anys", "Més de 20 anys"])) &
      (df_pd['Nom_Districte'] == 'Eixample')
     ]['Nombre'].sum() # Computing the sum of the 'Nombre' column

Wall time: 8.04 ms


75524

In [12]:
%%time
# Polars solution

# Filtering based on 'Antiguitat' and 'Nom_Districte'
df_pl.filter(
        (df_pl['Antiguitat'].is_in(pl.lit(["10 anys", "D'11 a 20 anys", "Més de 20 anys"]))) &
        (df_pl['Nom_Districte'] == 'Eixample')
            )['Nombre'].sum() # Computing the sum of the 'Nombre' column

Wall time: 8.01 ms


<timed eval>:5: DeprecationWarning: Behavior for `lit` will change for sequence inputs. The result will change to be a literal of type List. To retain the old behavior, pass a Series instead, e.g. `Series(sequence)`.


75524

**A:** In Eixample District, there were **75,524** vehicles that are older or 10 years old regostered in 2020.

### **Question #4**

What are the distinct neighbourhood names ('Nom_Barri') for the District ('Nom_Districte') 'Nou Barris'? (10 points)

In [13]:
%%time
# Pandas solution

# Filter DataFrame where 'Nom_Districte' is equal to 'Nou Barris'
df_pd['Nom_Barri'][df_pd['Nom_Districte'] == 'Nou Barris'] \
    .unique() # Unique 'Nom_Barri' values

Wall time: 0 ns


array(['Vilapicina i la Torre Llobeta', 'Porta', 'el Turó de la Peira',
       'Can Peguera', 'la Guineueta', 'Canyelles', 'les Roquetes',
       'Verdun', 'la Prosperitat', 'la Trinitat Nova', 'Torre Baró',
       'Ciutat Meridiana', 'Vallbona'], dtype=object)

In [14]:
%%time
# Polars Solution

# Filter DataFrame where 'Nom_Districte' is equal to 'Nou Barris'
df_pl.filter(df_pl['Nom_Districte'] == 'Nou Barris')['Nom_Barri'] \
    .unique()


Wall time: 0 ns


shape: (13,)
Series: 'Nom_Barri' [str]
[
	"Torre Baró"
	"Can Peguera"
	"la Guineueta"
	"Ciutat Meridia…
	"la Trinitat No…
	"Verdun"
	"Canyelles"
	"la Prosperitat…
	"Vilapicina i l…
	"el Turó de la …
	"Porta"
	"les Roquetes"
	"Vallbona"
]

### **Question #5**

What is the neighbourhood (‘Nom_Barri’) with more trucks (‘Camions’)? And how many trucks does it have? Look at the documentation for [function groupby in Pandas](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.groupby.html), parameter `as_index` to get the result as a dataframe instead of a series. (15 points)

In [15]:
%%time
# Pandas solution

# Filter DataFrame where 'Tipus_Vehicles' is 'Camions', then group by 'Nom_Barri', and sum 'Nombre'
df_pd_filtered_grouped = df_pd[df_pd['Tipus_Vehicles'] == 'Camions'] \
                                .groupby('Nom_Barri', as_index=False)['Nombre'] \
                                .sum()

# Get the neighborhood with the maximum 'Nombre'
df_pd_filtered_grouped.loc[[df_pd_filtered_grouped['Nombre'].idxmax()]]


Wall time: 11.4 ms


Nom_Barri  Nombre
57  la Marina del Prat Vermell - AEI Zona Franca    2059

In [16]:
%%time
# Polars Solution

# Filter DataFrame where 'Tipus_Vehicles' is 'Camions', then group by 'Nom_Barri', and sum 'Nombre'
df_pl_filtered_grouped = df_pl.filter(df_pl['Tipus_Vehicles'] == 'Camions') \
                              .group_by('Nom_Barri') \
                              .agg(pl.col('Nombre').sum().alias('Nombre'))

# Get the row with the maximum 'Nombre'
print(df_pl_filtered_grouped.sort(pl.col('Nombre'), descending=True) \
                      .slice(0, 1))  # This will get the first row of the sorted DataFrame


shape: (1, 2)
┌───────────────────────────────────┬────────┐
│ Nom_Barri                         ┆ Nombre │
│ ---                               ┆ ---    │
│ str                               ┆ i64    │
╞═══════════════════════════════════╪════════╡
│ la Marina del Prat Vermell - AEI… ┆ 2059   │
└───────────────────────────────────┴────────┘
Wall time: 40 ms


### **Question #6**
Rewrite the following pandas code snippet without the unction isin() (just & -AND-, | -OR- logical operators in the filter) (hint: be careful with parenthesis) (15 points).

In [17]:
df_pd[(df_pd['Antiguitat'].isin(["10 anys", "D'11 a 20 anys", "Més de 20 anys"])) & (df_pd['Tipus_Vehicles'] == "Turismes")]

Any Codi_Districte Nom_Districte Codi_Barri            Nom_Barri  \
10     2020              1  Ciutat Vella          1             el Raval   
11     2020              1  Ciutat Vella          1             el Raval   
12     2020              1  Ciutat Vella          1             el Raval   
72     2020              1  Ciutat Vella          1             el Raval   
73     2020              1  Ciutat Vella          1             el Raval   
...     ...            ...           ...        ...                  ...   
52896  2020             10    Sant Martí         73  la Verneda i la Pau   
52897  2020             10    Sant Martí         73  la Verneda i la Pau   
52946  2020             NC     No consta         NC            No consta   
52947  2020             NC     No consta         NC            No consta   
52948  2020             NC     No consta         NC            No consta   

      Seccio_Censal Tipus_Vehicles      Antiguitat  Nombre  
10                1       Turismes         10 anys      15  
11                1       Turismes  D'11 a 20 anys     132  
12                1       Turismes  Més de 20 anys      70  
72                2       Turismes         10 anys       5  
73                2       Turismes  D'11 a 20 anys      90  
...             ...            ...             ...     ...  
52896           143       Turismes  D'11 a 20 anys     242  
52897           143       Turismes  Més de 20 anys      51  
52946            NC       Turismes         10 anys       2  
52947            NC       Turismes  D'11 a 20 anys      56  
52948            NC       Turismes  Més de 20 anys      42  

[3207 rows x 9 columns]

In [18]:
# Code snippet without the isin()
df_pd[
        ((df_pd['Antiguitat'] == "10 anys") |
         (df_pd['Antiguitat'] == "D'11 a 20 anys") |
         (df_pd['Antiguitat'] == "Més de 20 anys")
        ) &
        (df_pd['Tipus_Vehicles'] == "Turismes")
     ]

Any Codi_Districte Nom_Districte Codi_Barri            Nom_Barri  \
10     2020              1  Ciutat Vella          1             el Raval   
11     2020              1  Ciutat Vella          1             el Raval   
12     2020              1  Ciutat Vella          1             el Raval   
72     2020              1  Ciutat Vella          1             el Raval   
73     2020              1  Ciutat Vella          1             el Raval   
...     ...            ...           ...        ...                  ...   
52896  2020             10    Sant Martí         73  la Verneda i la Pau   
52897  2020             10    Sant Martí         73  la Verneda i la Pau   
52946  2020             NC     No consta         NC            No consta   
52947  2020             NC     No consta         NC            No consta   
52948  2020             NC     No consta         NC            No consta   

      Seccio_Censal Tipus_Vehicles      Antiguitat  Nombre  
10                1       Turismes         10 anys      15  
11                1       Turismes  D'11 a 20 anys     132  
12                1       Turismes  Més de 20 anys      70  
72                2       Turismes         10 anys       5  
73                2       Turismes  D'11 a 20 anys      90  
...             ...            ...             ...     ...  
52896           143       Turismes  D'11 a 20 anys     242  
52897           143       Turismes  Més de 20 anys      51  
52946            NC       Turismes         10 anys       2  
52947            NC       Turismes  D'11 a 20 anys      56  
52948            NC       Turismes  Més de 20 anys      42  

[3207 rows x 9 columns]

### **Question #7**

Answer the question: what are the 5 (only 5) neighbourhoods (nom_barri) with more cars (Tipus_Vehicles equal Turismes) of age 20 years or more (Antiguitat equals 'Més de 20 anys'). You need to group by nom_barri, aggregating by Nombre and sorting by Nombre descending and keeping only the first 5 rows. Remember the as_index in the groupby or yow will have trouble sorting. (20 points)

In [19]:
%%time
# Pandas solution

# Filter DataFrame where 'Tipus_Vehicles' is 'Turismes' and 'Antiguitat' is 20 years or more, 
# then group by 'Nom_Barri', and sum 'Nombre'
df_pd[
      (df_pd['Antiguitat'].isin(["D'11 a 20 anys", "Més de 20 anys"])) & 
      (df_pd['Tipus_Vehicles'] == "Turismes")
     ] \
      .groupby('Nom_Barri', as_index=False)['Nombre'] \
      .sum() \
      .head(5) # Dsiplaying the 5 first records

Wall time: 8.11 ms


Nom_Barri  Nombre
0     Baró de Viver     389
1          Can Baró    1405
2       Can Peguera     337
3         Canyelles    1303
4  Ciutat Meridiana    1608

In [20]:
%%time
# Polars solution

# Filtering the DataFrame where 'Tipus_Vehicles' is 'Turismes' and 'Antiguitat' is 20 years or more
# then group by 'Nom_Barri', and sum 'Nombre'
print(df_pl.filter(
                (df_pl['Tipus_Vehicles'] == 'Turismes') & 
                ((df_pl['Antiguitat'] == "D'11 a 20 anys") | (df_pl['Antiguitat'] == "Més de 20 anys"))
            ) \
                .group_by('Nom_Barri') \
                .agg(pl.col('Nombre').sum().alias('Nombre')) \
                .slice(0, 5)) # Displaying the first 5 records


shape: (5, 2)
┌───────────────────────────────────┬────────┐
│ Nom_Barri                         ┆ Nombre │
│ ---                               ┆ ---    │
│ str                               ┆ i64    │
╞═══════════════════════════════════╪════════╡
│ el Parc i la Llacuna del Pobleno… ┆ 2254   │
│ el Poble Sec - AEI Parc Montjuïc  ┆ 4697   │
│ la Nova Esquerra de l'Eixample    ┆ 8337   │
│ Sant Gervasi - Galvany            ┆ 9731   │
│ la Verneda i la Pau               ┆ 4656   │
└───────────────────────────────────┴────────┘
Wall time: 41.3 ms


### **Final Note on Pandas vs Polars Performance**

Upon examining the execution times recorded in each exercise, an intriguing observation emerges: Polars consistently exhibits extended execution times compared to Pandas. This raises a question - is Polars inherently slower than Pandas?

There could be several factors contributing to this observed disparity in performance:

##### 1. Data Size:
   - **Small Data:** Polars is optimized for performance on large datasets. For smaller datasets, the overhead of optimizations can actually make Polars slower than pandas.

##### 2. Optimizations:
   - **Algorithms and Optimizations:** The specific algorithms and optimizations used by Polars might not always be as efficient for every type of operation or data. Some operations might be highly optimized in pandas after years of development and contributions.

##### 3. Multi-threading:
   - **Concurrency Model:** Polars uses a multi-threaded model which should make it faster for large datasets, but for smaller ones, the overhead of thread management could make it slower.


##### 4. Specific Operations:
   - **Operation Types:** Some operations might be optimized better in one library over the other. It can depend heavily on the specific use case or operation being performed.

##### 5. Implementation Language:
   - **Python vs Rust:** Polars is written in Rust, which is generally faster than Python. However, the interfacing between Python and Rust (when using Polars from Python) can sometimes introduce overhead.

### **Personal Reflection and Future Exploration**

I found this initial comparison between the Polars and Pandas libraries quite enlightening and I'm eager to extend this exploration with larger datasets for a more thorough analysis. In this context, Polars seemed somewhat overcooked, particularly considering its less extensive documentation and community support compared to Pandas, which can make it slightly more challenging to navigate.

However, on a positive note, I was drawn to the syntax of Polars for its resemblance to SQL commands, adding an intuitive aspect to its application. Despite the current limitations, there’s an evident potential in Polars that piques my curiosity. 

As I move forward, a detailed analysis utilizing more complex and larger datasets will provide clearer insights into the strengths and weaknesses of each library, offering a comprehensive view of their practical applicability and performance efficiency.